File: core/src/com/darkoverlordofdata/entitas/demo/systems/CollisionSystem.kt
package com.darkoverlordofdata.entitas.demo.systems

/**
 * Entitas Generated Systems for com.darkoverlordofdata.entitas.demo
 *
 */

import com.darkoverlordofdata.entitas.*
import com.darkoverlordofdata.entitas.demo.*

class CollisionSystem(pool: Pool)
      : IExecuteSystem {

    val pool = pool
    val bullets = pool.getGroup(Matcher.Bullet)
    val enemies = pool.getGroup(Matcher.Enemy)
    val players = pool.getGroup(Matcher.Player)

    override fun execute() {
        for (bullet in bullets.entities) {
            for (enemy in enemies.entities) {
                if (collidesWith(bullet, enemy))
                        collisionHandler(bullet, enemy)
            }
        }
    }

    fun collidesWith(e1:Entity, e2:Entity):Boolean {
        val position1 = e1.position
        var position2 = e2.position
        val a = (position1.x - position2.x).toDouble()
        val b = (position1.y - position2.y).toDouble()

        return ((Math.sqrt(a * a + b * b)) - e1.bounds.radius) < e2.bounds.radius
    }

    fun collisionHandler(weapon:Entity, ship:Entity) {
        val pos = weapon.position
        pool.createSmallExplosion(pos.x, pos.y)
        weapon.setDestroy(true)
        var health = ship.health
        health.currentHealth-=1
        if (health.currentHealth <= 0f) {
            val position = ship.position
            pool.createBigExplosion(position.x, position.y)
            ship.setDestroy(true)
            val player = players.singleEntity
            if (player != null)
                player.score.value += health.maximumHealth.toInt()
        }
    }
}

File: lib/src/systems/CollisionSystem.ts
module com.darkoverlordofdata.entitas.demo {

  import Pool = entitas.Pool;
  import Group = entitas.Group;
  import Entity = entitas.Entity;
  import Matcher = entitas.Matcher;
  import Exception = entitas.Exception;
  import TriggerOnEvent = entitas.TriggerOnEvent;
  import IInitializeSystem = entitas.IInitializeSystem;
  import IExecuteSystem = entitas.IExecuteSystem;
  import ISetPool = entitas.ISetPool;

  export class CollisionSystem implements IInitializeSystem, IExecuteSystem, ISetPool {

    protected pool:Pool;
    protected group:Group;

    public initialize() {
    }
    
    public execute() {
      var entities = this.group.getEntities();
      for (var i = 0, l = entities.length; i < l; i++) {
        var e = entities[i];
      }
    }
    
    public setPool(pool:Pool) {
      this.pool = pool;
      this.group = pool.getGroup(Matcher.allOf());
    }
    


  }
}